module/zfs/metaslab.c

```diff
@@ -5845,6 +5855,19 @@ metaslab_alloc(spa_t *spa, metaslab_class_t *mc, uint64_t psize, blkptr_t *bp,
     int ndvas, uint64_t txg, blkptr_t *hintbp, int flags,
     zio_alloc_list_t *zal, zio_t *zio, int allocator)
 {
+	if (mc->mc_ops->msop_type == METASLAB_TYPE_VIRTUAL) {
+		ASSERT3P(mc->mc_virtual, !=, NULL);
+		spa_t *target_spa = mc->mc_virtual;
+		dmu_tx_t *tx = dmu_tx_create_mos(target_spa->spa_dsl_pool);
+		VERIFY0(dmu_tx_assign(tx, TXG_WAIT | TXG_NOTHROTTLE));
+		uint64_t target_txg = dmu_tx_get_txg(tx);
+		int ret = metaslab_alloc(target_spa,
+		    spa_normal_class(target_spa), psize, bp, ndvas, target_txg,
+		    hintbp, flags, zal, zio, allocator);
+		dmu_tx_commit(tx);
```

Comment:
You are allocating space in transaction of different pool and without storing the pointer anywhere within that transaction.  What protect us from a space leak here if the transaction get committed immediately and we crash just here before we do anything else?

Reply:
Good catch.  I think the fix is to not commit the tx until the LWB's IOs are done and the new BP is safely on disk. This requires either having a thread block on the `zl_lwb_io_cv` or adding something to `zil_lwb_flush_vdevs_done` to commit the `tx`s.